### PR TITLE
Job and MPMD Job Implementation

### PR DESCRIPTION
**Pull request by juliaputko**

### CODE REVIEW

File: smartsim/launchable/job.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    LaunchSettings prior to launch.  It is responsible for turning
+    the stored entity and launch settings into commands that can be
+    executed by a launcher.
+
+    Jobs will hold a deep copy of launch settings.
+    """
+
+    def __init__(
+        self,
+        entity: SmartSimEntity,
```
[the remark this answers is not in the document]
Comment (amandarichardsonn) [reply smartsim/launchable/job.py:45]: I would also suggest assigning a deepcopy to the private attributes, just so it doesnt get all weird if a user edits an entity after creating the Job

File: smartsim/launchable/mpmdpair.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from smartsim.entity.entity import SmartSimEntity
+from smartsim.settings.base import RunSettings
+
+
+class MPMDPair:
```
Comment (amandarichardsonn) [inline smartsim/launchable/mpmdpair.py:31]: I think a dunder str might need to be added?

File: smartsim/launchable/mpmdjob.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        flag = 1
+
+
+class MPMDJob(BaseJob):
+    """An MPMDJob holds references to SmartSimEntity and
+    LaunchSettings pairs.  It is responsible for turning
+    The stored pairs into an MPMD command(s)
+    """
+
+    def __init__(self, mpmd_pairs: t.List[MPMDPair] = []) -> None:
```
Comment (amandarichardsonn) [inline smartsim/launchable/mpmdjob.py:74]: I would make the same edits to MPMDJob as Job! as far a public to private attr and public prop + deepcopy

**Review by amandarichardsonn — CHANGES_REQUESTED** [review]
Super awesome code!! Heres some thoughts/feedback

File: smartsim/launchable/mpmdpair.py (hunk 2 above)
Comment (amandarichardsonn) [inline smartsim/launchable/mpmdpair.py:31]: Does this class inherit from another?

File: tests/test_launchable.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+def test_launchable_init():
+    launchable = Launchable()
+    assert isinstance(launchable, Launchable)
+
+
+def test_job_init():
+    entity = SmartSimEntity("test_name", None, None)
+    launch_settings = RunSettings("echo", ["spam", "eggs"])
+    job = Job(entity, launch_settings)
```
Comment (amandarichardsonn) [inline tests/test_launchable.py:53]: Could you assert that these values were set correctly? Also once you add the deepcopy, could you make a change to the original address to make sure and make sure the attribute doesnt change?

File: tests/test_launchable.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    using add_mpmd_pair"""
+    entity = SmartSimEntity("test_name", None, None)
+    launch_settings = RunSettings("echo", ["spam", "eggs"])
+
+    mpmd_job = MPMDJob()
+    mpmd_job.add_mpmd_pair(entity, launch_settings)
+
+    mpmd_pair = MPMDPair(entity, launch_settings)
+
+    assert len(mpmd_job.mpmd_pairs) > 0
```
Comment (amandarichardsonn) [inline tests/test_launchable.py:68]: could you instead make sure the length is the exact expected value?

File: tests/test_launchable.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    assert mpmd_pair.entity == mpmd_job.mpmd_pairs[0].entity
+    assert mpmd_pair.launch_settings == mpmd_job.mpmd_pairs[0].launch_settings
+
+
+def test_mpmdpair_init():
+    """Test the creation of an MPMDPair"""
+    entity = SmartSimEntity("test_name", None, None)
```
Comment (amandarichardsonn) [inline tests/test_launchable.py:76]: check the attributes! And do the deepcopy check!

File: smartsim/launchable/mpmdpair.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    """Class to store MPMD Pairs"""
+
+    def __init__(
+        self, entity: SmartSimEntity, launch_settings: RunSettings
+    ):  # rename to LaunchSettings
+        self.entity = entity
+        self.launch_settings = launch_settings
+
+    def __iter__(self):
```
Comment (amandarichardsonn) [inline smartsim/launchable/mpmdpair.py:40]: what is this __iter__ for? It seems like MPMDPair is just comprised of entity and launch_settings? Comment for understanding and my learnin!

File: smartsim/launchable/mpmdjob.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                print(type(mpmd_pair.entity))
+            else:
+                raise SSUnsupportedError(
+                    "MPMD pairs must all share the same entity type."
+                )
+        ret = mpmd_pair.entity
+        flag = 1
+
+
+class MPMDJob(BaseJob):
```
Comment (amandarichardsonn) [inline smartsim/launchable/mpmdjob.py:68]: I would make MPMD into a Collection container! So it behaves like JobGroup or ColocatedJobGroup. You can do this by import `collections.abc.Collections` and inheriting Collections. And then you'll just need to add the dunder methods __len__, __iter__ and __contains__, what do you think?

File: smartsim/launchable/mpmdjob.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+def _check_entity(mpmd_pairs: t.List[MPMDPair]) -> None:
+    """Enforce all pairs have the same entity types"""
+    flag = 0
+    ret = None
+    for mpmd_pair in mpmd_pairs:
+        if flag == 1:
+            if type(ret) == type(mpmd_pair.entity):
+                flag = 0
+                print(type(ret))
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/launchable/mpmdjob.py:58]: removed

File: smartsim/_core/_install/builder.py [hunk 10]
```diff
@@ -705,8 +705,9 @@ def _install_backends(self, device: Device) -> None:
         rai_lib = self.rai_install_path / "redisai.so"
         rai_backends = self.rai_install_path / "backends"
 
-        if rai_lib.is_file() and rai_backends.is_dir():
+        if rai_backends.is_dir():
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/_core/_install/builder.py:708]: I did not!

File: smartsim/launchable/basejob.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+from abc import ABC, abstractmethod
+
+from smartsim.launchable.launchable import Launchable
+
+
+class BaseJob(ABC, Launchable):
+    """The highest level abstract base class for a single job that can be launched"""
+
+    @abstractmethod
+    def get_launch_steps(self) -> None:  # -> LaunchSteps:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/launchable/basejob.py:37]: added

File: smartsim/launchable/job.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    def get_launch_steps(self) -> None:  # -> LaunchCommands:
+        """Return the launch steps corresponding to the
+        internal data.
+        """
+        pass
+        # return JobWarehouseRunner.run(self)
+
+    def __str__(self) -> str:
+        string = f"SmartSim Entity: {self.entity}"
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/launchable/job.py:61]: added

File: smartsim/launchable/job.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        # self.warehouse_runner = JobWarehouseRunner
+
+    def get_launch_steps(self) -> None:  # -> LaunchCommands:
+        """Return the launch steps corresponding to the
+        internal data.
+        """
+        pass
+        # return JobWarehouseRunner.run(self)
+
+    def __str__(self) -> str:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/launchable/job.py:60]: added

File: smartsim/launchable/mpmdjob.py [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+class MPMDJob(BaseJob):
+    """An MPMDJob holds references to SmartSimEntity and
+    LaunchSettings pairs.  It is responsible for turning
+    The stored pairs into an MPMD command(s)
+    """
+
+    def __init__(self, mpmd_pairs: t.List[MPMDPair] = []) -> None:
+        super().__init__()
+
+        self.mpmd_pairs = mpmd_pairs
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/launchable/mpmdjob.py:77]: Changed to your suggestions

File: smartsim/launchable/job.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/launchable/job.py:45]: Changed to suggestion

File: smartsim/launchable/mpmdpair.py (hunk 2 above)
Comment (juliaputko) [reply smartsim/launchable/mpmdpair.py:31]: added

File: smartsim/launchable/mpmdjob.py (hunk 3 above)
Comment (juliaputko) [reply smartsim/launchable/mpmdjob.py:74]: like this? self.mpmd_pairs = deepcopy(mpmd_pairs) or []

File: smartsim/launchable/mpmdpair.py (hunk 2 above)
Comment (juliaputko) [reply smartsim/launchable/mpmdpair.py:31]: I dont think so! The purpose of the class is for easier access to the MPMDPairs. Basically easier access than from a tuple

File: smartsim/launchable/mpmdpair.py (hunk 7 above)
Comment (juliaputko) [reply smartsim/launchable/mpmdpair.py:40]: To avoid tuples!

File: tests/test_launchable.py (hunk 5 above)
Comment (juliaputko) [reply tests/test_launchable.py:68]: todo

File: tests/test_launchable.py (hunk 4 above)
Comment (juliaputko) [reply tests/test_launchable.py:53]: added

File: smartsim/launchable/mpmdjob.py (hunk 3 above)
Comment (amandarichardsonn) [reply smartsim/launchable/mpmdjob.py]: self.mpmd_pairs = copy.deepcopy(mpmd_pairs) if mpmd_pairs else []

**Review by amandarichardsonn — APPROVED** [review]
LGTM!

File: smartsim/launchable/mpmdjob.py (hunk 3 above)
Comment (juliaputko) [reply smartsim/launchable/mpmdjob.py]: changed

**Pull request merged**